airflow-core/src/airflow/ui/tests/e2e/pages/BackfillPage.ts

```
@@ -0,0 +1,209 @@
+
+    const runsMessage = this.page.locator("text=/\\d+ runs? will be triggered|No runs matching/");
+
+    await expect(runsMessage).toBeVisible({ timeout: 10_000 });
+
+    await expect(this.backfillRunButton).toBeEnabled({ timeout: 5000 });
+    await this.backfillRunButton.click();
+  }
+
+  // Get backfill details from a specific row (default: first row)
+  public async getBackfillDetails(rowIndex: number = 0): Promise<{
+    createdAt: string;
+    fromDate: string;
+    reprocessBehavior: string;
+    toDate: string;
+  }> {
+    const row = this.page.locator("table tbody tr").nth(rowIndex);
+    const cells = row.locator("td");
+
+    const fromDate = (await cells.nth(0).textContent()) ?? "";
+    const toDate = (await cells.nth(1).textContent()) ?? "";
+    const reprocessBehavior = (await cells.nth(2).textContent()) ?? "";
+    const createdAt = (await cells.nth(3).textContent()) ?? "";
+
+    return {
+      createdAt: createdAt.trim(),
+      fromDate: fromDate.trim(),
+      reprocessBehavior: reprocessBehavior.trim(),
+      toDate: toDate.trim(),
+    };
+  }
+
+  public async getBackfillsTableRows(): Promise<number> {
+    const rows = this.page.locator("table tbody tr");
+
+    await rows.first().waitFor({ state: "visible", timeout: 10_000 });
+    const count = await rows.count();
+
+    return count;
+  }
+
+  // Get filter button
+  public getFilterButton(): Locator {
+    return this.page.locator('button[aria-label*="filter"], button[aria-label*="Filter"]');
+  }
+
+  // Get number of table columns
+  public async getTableColumnCount(): Promise<number> {
+    const headers = this.page.locator("table thead th");
+
+    return await headers.count();
+  }
+
+  public async isBackfillDateErrorVisible(): Promise<boolean> {
+    return this.backfillDateError.isVisible();
+  }
+
+  //  Check if a specific column is visible in the table
+  public async isColumnVisible(columnName: string): Promise<boolean> {
```

ColumnVisible() and isFilterAvailable() catch errors and return false. This makes debugging difficult
1. Test failures show unhelpful "Expected: true, Received: false" messages
2. You don't know if the element is missing, selector is wrong, or page didn't load
3. Combined with if conditions, tests can silently pass without testing anything